File: src/main/java/com/rh360/rh360/service/AuthService.java
package com.rh360.rh360.service;

import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import com.rh360.rh360.dto.LoginRequest;
import com.rh360.rh360.dto.LoginResponse;
import com.rh360.rh360.entity.User;
import com.rh360.rh360.repository.UsersRepository;

@Service
public class AuthService {

    private final UsersRepository usersRepository;
    private final TokenService tokenService;
    private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();

    public AuthService(UsersRepository usersRepository, TokenService tokenService) {
        this.usersRepository = usersRepository;
        this.tokenService = tokenService;
    }

    public LoginResponse login(LoginRequest request) {
        User user = usersRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Usuário não encontrado"));

        if (!"active".equals(user.getStatus())) {
            throw new RuntimeException("Usuário inativo");
        }

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new RuntimeException("Senha incorreta");
        }

        // Gerar token JWT
        String token = tokenService.generateToken(
                user.getId(),
                user.getEmail(),
                user.getRole() != null ? user.getRole() : "USER"
        );

        // Salvar token no banco de dados
        tokenService.saveToken(token, user.getId());

        return new LoginResponse(user.getId(), token);
    }

    public void logout(String token) {
        tokenService.deactivateToken(token);
    }
}

File: src/auth/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import * as bcrypt from 'bcrypt';
import { User } from '../entities/user.entity';
import { TokenService } from '../token/token.service';
import { LoginRequestDto } from './dto/login-request.dto';
import { LoginResponseDto } from './dto/login-response.dto';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private usersRepository: Repository<User>,
    private tokenService: TokenService,
    private jwtService: JwtService,
  ) {}

  async login(loginDto: LoginRequestDto): Promise<LoginResponseDto> {
    const user = await this.usersRepository.findOne({
      where: { email: loginDto.email },
    });

    if (!user) {
      throw new UnauthorizedException('Usuário não encontrado');
    }

    if (user.status !== 'active') {
      throw new UnauthorizedException('Usuário inativo');
    }

    const isPasswordValid = await bcrypt.compare(
      loginDto.password,
      user.password,
    );

    if (!isPasswordValid) {
      throw new UnauthorizedException('Email ou Senha incorreta');
    }

    // Invalidar todos os tokens anteriores do usuário
    await this.tokenService.deactivateAllUserTokens(user.id);

    // Gerar novo token JWT
    const token = await this.tokenService.generateToken(
      user.id,
      user.email,
      user.role || 'user',
    );

    // Salvar novo token no banco de dados
    await this.tokenService.saveToken(token, user.id);

    return {
      id: user.id,
      token,
    };
  }

  async logout(token: string): Promise<void> {
    await this.tokenService.deactivateToken(token);
  }
}
